### PR TITLE
Fix iSAM2 support for smart factors

### LINKED ISSUE #10: iSAM2 and smartfactors
Re-opening [bitbucket issue 420](https://bitbucket.org/gtborg/gtsam/issues/420/problem-with-isam2-stereo-smart-factors-no)

### PR DESCRIPTION
**Pull request by jlblancoc**
A new optional parameter has been added to iSAM2::update(), only needed when using smart factors, or other factors that depend on a dynamic number of Keys.

A new unit test has been added: it runs a toy Stereo SLAM problem with both, batch LM, and iSAM2. It has a "verbosity" macro that can be commented out to enable printing all the details. Originally, the test failed with an error of "ill posed problem", it works now with iSAM2 giving exactly the same result than LM.

Closes #10

### CODE REVIEW

File: gtsam/nonlinear/ISAM2.h [hunk 1]
```diff
@@ -156,6 +157,20 @@ class GTSAM_EXPORT ISAM2 : public BayesTree<ISAM2Clique> {
       const boost::optional<FastList<Key> >& extraReelimKeys = boost::none,
       bool force_relinearize = false);
 
+  /**
+   * Alternative signature of update() with all additional parameters in one
+   * structure. This form makes easier to keep future API/ABI compatibility if
+   * parameters change.
+   *
+   * @param extraParams Additional parameters to control relinearization,
```
Comment (dellaert) [inline gtsam/nonlinear/ISAM2.h]: out of date? Also add other params?

File: gtsam/nonlinear/ISAM2.h [hunk 2]
```diff
@@ -156,6 +157,20 @@ class GTSAM_EXPORT ISAM2 : public BayesTree<ISAM2Clique> {
       const boost::optional<FastList<Key> >& extraReelimKeys = boost::none,
       bool force_relinearize = false);
 
+  /**
+   * Alternative signature of update() with all additional parameters in one
```
Comment (dellaert) [inline gtsam/nonlinear/ISAM2.h]: I love it!

File: gtsam/nonlinear/ISAM2.h [hunk 3]
```diff
@@ -156,6 +157,20 @@ class GTSAM_EXPORT ISAM2 : public BayesTree<ISAM2Clique> {
...
+   * parameters change.
+   *
+   * @param extraParams Additional parameters to control relinearization,
+   * constrained keys, etc.
+   * @return An ISAM2Result struct containing information about the update
+   * @note No default parameters to avoid ambiguous call errors.
+   */
+  virtual ISAM2Result update(
+      const NonlinearFactorGraph& newFactors, const Values& newTheta,
+      const ISAM2UpdateParams& up);
```
Comment (dellaert) [inline gtsam/nonlinear/ISAM2.h]: suggest renaming `up` to `updateParams`

File: gtsam/nonlinear/ISAM2.cpp [hunk 4]
```diff
@@ -544,6 +544,22 @@ ISAM2Result ISAM2::update(
...
+  params.removeFactorIndices = removeFactorIndices;
+
+  return update(newFactors, newTheta, params);
+}
+
+/* ************************************************************************* */
+ISAM2Result ISAM2::update(
+    const NonlinearFactorGraph& newFactors,
+    const Values& newTheta,
+    const ISAM2UpdateParams& up) {
```
Comment (dellaert) [inline gtsam/nonlinear/ISAM2.cpp]: suggest renaming `up` to `updateParams`

File: gtsam/nonlinear/ISAM2.cpp [hunk 5]
```diff
@@ -649,28 +666,34 @@ ISAM2Result ISAM2::update(
...
-    for (Key key : *extraReelimKeys) {
+  if (up.extraReelimKeys) {
+    for (Key key : *up.extraReelimKeys) {
       markedKeys.insert(key);
     }
   }
+  // Also, keys that were not observed in existing factors, but whose affected
+  // keys have been extended now (e.g. smart factors)
+  if (up.newAffectedKeys) {
+    for (const auto &f2ks : up.newAffectedKeys.value()) {
```
Comment (dellaert) [inline gtsam/nonlinear/ISAM2.cpp]: use different variable. f2ks ??

File: gtsam/nonlinear/ISAM2UpdateParams.h [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+   * reordered. */
+  boost::optional<FastList<Key>> extraReelimKeys{boost::none};
+
+  /* Relinearize any variables whose delta magnitude is sufficiently large
+   * (Params::relinearizeThreshold), regardless of the relinearization
+   * interval (Params::relinearizeSkip). */
+  bool force_relinearize{false};
+
+  /** An optional set of new Keys that are now affected by factors,
+   * indexed by factor indices (as returned by ISAM2::update()).
```
Comment (dellaert) [inline gtsam/nonlinear/ISAM2UpdateParams.h]: Say motivation is smartfactors, perhaps illustrated with small example.

File: gtsam/nonlinear/ISAM2.cpp [hunk 7]
```diff
@@ -773,14 +796,22 @@ ISAM2Result ISAM2::update(
     variableIndex_.augment(newFactors, result.newFactorsIndices);
   else
     variableIndex_.augment(newFactors);
+
+  // Augment it with existing factors which now affect to more variables:
+  if (up.newAffectedKeys) {
+    for (const auto &fk : *up.newAffectedKeys) {
```
Comment (dellaert) [inline gtsam/nonlinear/ISAM2.cpp]: fk ? In general, don't abbreviate. I know we used to do this, but I frown upon it now :-)

File: gtsam/nonlinear/ISAM2.cpp [hunk 8]
```diff
@@ -773,14 +796,22 @@ ISAM2Result ISAM2::update(
...
+      const auto factorIdx = fk.first;
+      variableIndex_.augmentExistingFactor(factorIdx, fk.second);
```
Comment (dellaert) [inline gtsam/nonlinear/ISAM2.cpp]: Very scared of this imperative programming construct, but then again, iSAM2 is already scarily imperative. best we can hope for is we have good unit tests and did not forget about scary unforeseen side effects.

**Comment by jlblancoc** [comment]
Oh, abbreviations, I forgot :-)

All [done](https://github.com/borglab/gtsam/pull/25/commits/d6b24294712db197096cd3ea75fbed3157aea096). `make check` is happy.

**Comment by dellaert** [comment]
Thanks for this excellent contribution!

**Pull request merged**